src/AWS.elm

```diff
@@ -4,31 +4,89 @@ module AWS exposing (..)
 
 __Experimental: Work in progress__
 
-@docs Credentials, ServiceConfig
+@docs ServiceConfig, Credentials, credentials, AccessKeyId, SecretAccessKey, SessionToken
+
+## Request Signing
+
+@docs Request, SignedRequest, signV4
 
 -}
 
+import AWS.Config
+import AWS.Signers.V4 exposing (sign)
+import AWS.Http
+import Date exposing (Date)
+import Http
+
 
 {-| Credentials for accessing AWS services.
 -}
-type alias Credentials =
-    { accessKeyId : String
-    , secretAccessKey : String
-    , sessionToken : Maybe String
-    }
+type Credentials
+    = Credentials AWS.Config.Credentials
+
+
+{-| The AWS access key ID.
+-}
+type alias AccessKeyId =
+    String
+
+
+{-| The AWS secret access key.
+-}
+type alias SecretAccessKey =
+    String
+
+
+{-| An optional AWS session token.
+-}
+type alias SessionToken =
+    String
+
+
+{-| Constructor for AWS Credentials
+-}
+credentials : AccessKeyId -> SecretAccessKey -> Maybe SessionToken -> Credentials
+credentials accessKeyId secretAccessKey maybeToken =
+    AWS.Config.Credentials accessKeyId secretAccessKey maybeToken
+        |> Credentials
 
 
 {-| Configuration for an AWS service.
 
-Each service module exposes a `config` function which is preferable to
-constructing this record type directly.
--}
-type alias ServiceConfig =
-    { serviceName : String
-    , version : String
-    , xAmzJsonVersion : String
-    , xAmzTargetPrefix : String
-    , host : String
-    , region : String
-    , credentials : Maybe Credentials
-    }
+Each service module exposes a `config` function which returns a ServiceConfig
+for use with that service.
+-}
+type ServiceConfig
+    = ServiceConfig AWS.Config.Service
+
+
+{-| An unsigned AWS Request.
+
+Call signV4 to produce a signed Http.Request
+-}
+type Request a
+    = UnsignedRequest (AWS.Http.UnsignedRequest a)
+
+
+{-| Alias for Http.Request
+
+See http://package.elm-lang.org/packages/elm-lang/http/1.0.0/Http#Request
+-}
+type alias SignedRequest a =
+    Http.Request a
+
+
+{-| Sign an unsigned request, given the current time
+-}
+signV4 :
+    ServiceConfig
+    -> Credentials
+    -> Date
+    -> Request a
+    -> Result String (SignedRequest a)
+signV4 serviceConfig credentials date req =
+    case
+        ( serviceConfig, credentials, req )
+    of
+        ( ServiceConfig config, Credentials creds, UnsignedRequest unsignedReq ) ->
```

Comment:
exposed function unwraps from the opaque types and forwards to the real signing function